436 - Add support for legacy compendia routes

nozomione opened the pull request:
## Issue Number

Closes #436

## Purpose/Implementation Notes

(resource) [Next.js Docs - NextResponse.redirect in Middleware](https://nextjs.org/docs/pages/building-your-application/routing/redirecting#nextresponseredirect-in-middleware)

I've added support for legacy `compendia` routes using Next.js `/src/middleware`.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests

N/A

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

davidsmejia reviewed (commented):
Since we cant infer the type of compendium from the old download url, lets just redirect to the compendia landing page.

Instead of creating two components this can be acheived with 3 redirect rule in next.config.js.

1. when c=normalized -> /normalized
2. when c=rna-seq -> /rna-seq
3. /compendia/download -> /compendia

Docs for this can be found here: https://nextjs.org/docs/app/api-reference/config/next-config-js/redirects#header-cookie-and-query-matching

nozomione commented:
> Since we cant infer the type of compendium from the old download url, lets just redirect to the compendia landing page.
> 
> Also, the download link expires anyway so we can assume that they aren't being saved.

I've updated the redirects logic to use Next.js `middleware` instead of `next.config`, as I discovered that `redirects` in `next.config` appends legacy query parameters (e.g.,`c`) to the `destination` URLs (e.g., `/compendia/normalized?c=normalized`), and there's no direct way to exclude them. Additionally, I've updated the issue to reflect this change and included a link to the middleware doc page. 

This PR is ready for your review. Thank you, David!

davidsmejia reviewed (commented):
You need to add a "matcher" to only apply this middleware to compedia routes. Ideally, it would be one middleware per type of redirect. 

Where one middleware redirects to the correct tab, the other redirects away from download_url in query params.

Alternatively, 
You can use next.config.js and redirect, then on the redirected page, detect query params and redirect to the same page without the extraneous queries.

nozomione commented:
I've used the first approach and included the matcher with specified paths. 

This PR is ready for another look. Thank you, David!

davidsmejia reviewed (approved):
LGTM

nozomione commented:
@avrohomgottlieb This PR is ready for your review. Ping me if you have any questions. Thank you!
